### PR TITLE
[4.2] Inject the cache controller factory

### PR DESCRIPTION
**Pull request by laoneo**
### Summary of Changes
Adds a new interface and trait to inject a `CacheControllerFactoryInterface`. Where possible this is used now in core.

### Testing Instructions
Open the back end.

### Actual result BEFORE applying this Pull Request
All works.

### Expected result AFTER applying this Pull Request
All works.

### CODE REVIEW

**Comment by brianteeman** [comment]
Impossible to test from the instructions provided

**Comment by SumCompanyInc** [comment]
the backend works after applying the patch?<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/37517">issues.joomla.org/tracker/joomla-cms/37517</a>.</sub>

**Comment by SumCompanyInc** [comment]
I have tested this item :white_check_mark: successfully on 7205827bb3f3e7dce4539af24c7bdf6afc424371<br /><br />if it's just opening the backend, it opened fine for me after applying patch<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/37517">issues.joomla.org/tracker/joomla-cms/37517</a>.</sub>

**Comment by joomdonation** [comment]
I have tested this item :white_check_mark: successfully on bb90f76dd63b6e7b33d404e2d3e0fa7bb6fcb5c7<br /><br />Apply patch, access to random pages from both frontend and backend, with both cache enabled/disabled, nothing is broken.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/37517">issues.joomla.org/tracker/joomla-cms/37517</a>.</sub>

**Comment by joomdonation** [comment]
@SumCompanyInc Could you please re-test this PR ? Thanks.

**Comment by laoneo** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/37517">issues.joomla.org/tracker/joomla-cms/37517</a>.</sub>

**Comment by laoneo** [comment]
No need for a second test as I added only a unit test and made a function protected.

File: libraries/src/MVC/Model/BaseDatabaseModel.php [hunk 1]
```diff
@@ -311,8 +313,8 @@ protected function cleanCache($group = null)
 
 		try
 		{
-			/** @var CallbackController $cache */
-			$cache = Factory::getContainer()->get(CacheControllerFactoryInterface::class)->createCacheController('callback', $options);
+			 /** @var CallbackController $cache */
```
Comment (Quy) [inline libraries/src/MVC/Model/BaseDatabaseModel.php]:
```suggestion
			/** @var CallbackController $cache */
```

**Comment by roland-d** [comment]
Thanks everybody

**Pull request merged**